src/core/net/ip6.cpp

```diff
@@ -1021,27 +1021,51 @@ const NetifUnicastAddress *Ip6::SelectSourceAddress(MessageInfo &aMessageInfo)
                     rvalPrefixMatched = candidatePrefixMatched;
                 }
             }
+            else if (rvalAddr->GetScope() == Address::kRealmLocalScope)
+            {
+                // Rule 3: Prefer appropriate realm local address
+                if (overrideScope > Address::kRealmLocalScope)
+                {
+                    if (rvalAddr->GetAddress().IsRoutingLocator())
+                    {
+                        // Prefer EID if destination is not realm local.
+                        rvalAddr = addr;
+                        rvalIface = candidateId;
+                        rvalPrefixMatched = candidatePrefixMatched;
+                    }
+                }
+                else
+                {
+                    if (candidateAddr->IsRoutingLocator())
+                    {
+                        // Prefer RLOC if destination is realm local.
+                        rvalAddr = addr;
+                        rvalIface = candidateId;
+                        rvalPrefixMatched = candidatePrefixMatched;
+                    }
+                }
+            }
             else if (addr->mPreferred && !rvalAddr->mPreferred)
             {
-                // Rule 3: Avoid deprecated addresses
+                // Rule 4: Avoid deprecated addresses
```

Comment:
These rule numbers correspond to [RFC 6724 Section 5](https://tools.ietf.org/html/rfc6724#section-5).